### ISSUE TITLE
Type error when using `ngFor` on value of type `number[] | string[]`

### ISSUE THREAD

**Issue by meblum** [angular/angular#43162]
### Which @angular/* package(s) are the source of the bug?

common

### Is this a regression?

No

### Description

When looping on a value of type `number[] | string[]` using `*ngFor`, I'm getting an error:
`Type 'string[] | number[]' is not assignable to type '(string[] & NgIterable<string>) | null | undefined'.ngtsc(2322)`

example code:

```
import { Component } from '@angular/core';
@Component({
  selector: 'my-app',
  template: `
    <div *ngFor="let v of val">
      {{ v }}
    </div>
  `,
  styles: ['']
})
export class AppComponent {
  val: number[] | string[] = [1, 2, 3, 4, 5, 6, 7];
}

```

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-ivy-xhzrjk?file=src%2Fapp%2Fapp.component.ts

### Please provide the exception or error you saw

```true
Type 'string[] | number[]' is not assignable to type '(string[] & NgIterable<string>) | null | undefined'.ngtsc(2322)
```

**Comment by JoostK** [comment]
This is also not allowed in TypeScript itself, [as seen in this playground example](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAhjAvDA2gRgDQwEwF14QzQBOAlmAOYr4A+MYArgLYBGApsdQNwCwAUPyhtoACjgBKXgL4AzBmGBRS4GEOgAeACoA+EQA8AXDGJs4AE3AAbAJ4xN1cTADe-fgF9+QA).

Only if `test` is invoked with an explicit type argument `<string|number>` does it compile, but then the array's type becomes `(string | number)[]` instead of `string[] | number[]`. In the context of `NgForOf` I believe that is still safe as `NgForOf` does not write to the array and only exposes individual values, but the types _are_ incompatible in the sense that `string[] | number[]` gives stronger guarantees about the contents of the array so TypeScript reports an error. I'm not sure if there would be a way for `NgForOf` to start accepting this pattern (or perhaps there is even a scenario where this is actually unsafe).

tldr; Use `(number|string)[]` as the type. It is not as strong as `number[] | string[]` but would be accepted by `NgForOf`.

**Comment by meblum** [comment]
I see. Thank you.

**Comment by meblum** [comment]
@JoostK This does work,
